src/classes/CRLP_RollupProcessor.cls

```diff
@@ -0,0 +1,538 @@
+            // collection of CRLP_Rollup instances in the stateful options var. The completeRollupForParent()
+            // methods stores everything in this var because it's not sure of which is the first or last.
+            // The block below removes everything from this map that is not the first summary record id.
+            if (itemNum > 1 && previousParentIdProcessed != null) {
+                rollupsDefsByParentId.remove(previousParentIdProcessed);
+            }
+
+            // Call the method to complete the actual rollup operation for this parent.
+            // The resulting SObject record is returned.
+            // The options.rollups collection will contain the list of CRLP_Rollup records
+            // that were processed for this rollup operation. This is needed to determine if the
+            // resulting record requires an update.
+            SObject updatedRecord = completeRollupForSingleSummaryRecord(parentRecords[itemNum]);
+
+            // Does the resulting record require an update?
+            if (updatedRecord != null) {
+                recordsToUpdate.put(parentId, updatedRecord);
+            }
+
+            if (parentListSize > 1) {
+                CRLP_Debug_UTIL.clearCurrentRollupState();
+            }
+            previousParentIdProcessed = parentId;
+        }
+
+        // Need to ensure the last record isn't removed it if needs to be written. Just not sure
+        // how to know that yet.
+
+        // When running in LDV mode, the last ParentId process should NOT be updated yet (in the batch iteration).
+        // This is necessary to handle the situation when a single Parent record may have thousands of detail records
+        // passed through multiple batch iterations. The 'final' record will be handled by the finish() method of
+        // the LDV Batch job.
+        lastParentIdRemovedFromUpdatesList = false;
+        if (isSkewMode
+                && lastParentIdProcessed != null
+                && recordsToUpdate.containsKey(lastParentIdProcessed)
+        ) {
+            recordsToUpdate.remove(lastParentIdProcessed);
+            lastParentIdRemovedFromUpdatesList = true;
+        }
+
+        return recordsToUpdate.values();
+    }
+
+    /**
+    * @description When PreventDupesInRollup is true, this method will purge the list of detail Opps in the detailsForParent
+    * collection to remove any that were previously processed. This is used in AccountContactSoftCredit rollups to
+    * prevent an Account from both getting soft credit for the same opportunity more than one - when there is more than
+    * one Household Contact that got soft credit for that Opportunity.
+    */
+    private List<SObject> cleanDetailOppsListForSoftCreditMode(Id parentId, List<SObject> detailsForParent) {
+        List<SObject> oppsToRollup = new List<SObject>();
+        for (Integer itemNum = 0, detailListSize = detailsForParent.size(); itemNum < detailListSize; itemNum++) {
```

Comment:
Lines 384-385 can be: `for (SObject detail : detailsForParent) {`

Reply:
The reason that these loops use an integer iterator over a Sobject is the result of performance analysis that originated here: https://paulbattisson.com/blog/2014/force-com-cast-episode-6-faster-loops/. At the time of building CRLP I ran tests of my own and confirmed that this is actually faster. If the loop was only performed once it probably wouldn't make any difference, but given the nature of CRLP many of these loops are performed hundreds or thousands times depending on the amount of data being rolled up.

Reply:
Thanks for letting me, I was not aware of the performance issue. If so, let's be consistent in naming and reduce the variable names throughout code, for example: `for (Integer i = 0, size = detailsForParent.size(); i < size; i++) {`? Sometimes the index is `n`,  `itemNum` or other, so let's consistently name vars in `CRLP` namespace?
Similarly `detailsForParent` are just `detailRecords`?
Or if we specifically know where are getting OCRs or PSCs, let's name those detail record vars and methods as such?

Reply:
Makes sense.